jax/_src/tree.py

```diff
@@ -284,3 +286,45 @@ def unflatten(treedef: tree_util.PyTreeDef,
     - :func:`jax.tree.structure`
   """
   return tree_util.tree_unflatten(treedef, leaves)
+
+
+def field(*, static: bool = False, metadata: dict[str, Any] | None = None, **kwds) -> dataclasses.Field:
+  """Used to mark a jax.dataclass field as static.
+
+  See {func}`jax.tree.dataclass` for details.
+  """
+  return dataclasses.field(metadata=(metadata or {}) | {'static': static}, **kwds)
+
+
+@dataclass_transform(field_specifiers=(field, dataclasses.Field), frozen_default=True)
+def dataclass(cls: T, /, **kwargs) -> T:
+  """Create a pytree-compatible dataclass.
+
+  Examples:
+    >>> import jax
+    >>> @jax.tree.dataclass
+    ... class OpDescription:
+    ...   op: str = jax.tree.field(static=True)
+    ...   val1: jax.Array
+    ...   val2: jax.Array
+    ...
+    >>> @jax.jit
+    ... def f(desc: OpDescription):
+    ...   if desc.op == "add":
+    ...     return desc.val1 + desc.val2
+    ...   else:
+    ...     raise NotImplementedError(f"{desc.op=}")
+    ...
+    >>> x = OpDescription("add", jnp.arange(4), 10)
+    >>> f(x)
+    Array([10, 11, 12, 13], dtype=int32)
+  """
+  if '_jax_dataclass' in cls.__dict__:
+    return cls
+  kwargs.setdefault('frozen', True)
+  dcls = dataclasses.dataclass(**kwargs)(cls)  # type: ignore
+  dcls._jax_dataclass = True
+  data_fields = [f.name for f in dataclasses.fields(dcls) if not f.metadata.get('static', False)]
```

Comment:
`dcls.replace = dataclasses.replace`  please please please

Reply:
Just to re-iterate my earlier argument, there is currently no way of telling Python type checkers that the result of `dataclass_transform` has an extra method. We need intersection types for that.

Reply:
One way we make type checkers happy in the flax version is by using a base type, roughly:

```python
TNode = TypeVar('TNode', bound='PyTreeNode')
@dataclass_transform(field_specifiers=(field,))  # type: ignore[literal-required]
class PyTreeNode:
  def __init_subclass__(cls, **kwargs):
    dataclass(cls, **kwargs)  # pytype: disable=wrong-arg-types
  def __init__(self, *args, **kwargs):
    # stub for pytype
    raise NotImplementedError
  def replace(self: TNode, **overrides) -> TNode:
    # stub for pytype
    raise NotImplementedError
```
one uses it like
```
class Thing(struct.PyTreeNode):
  blah: Any
  ...
```

I'd just like some convenience decorator or base class that does the common stuff we use _constantly_ in current implementations.  If we want the blessed interface to be stdlib compliant, fine - but I'd love something without a ton of syntactic boilerplate.

Reply:
Yeah, actually, I agree that a base class approach would work. I think we prototyped this for Flax at some point?